core/shared/src/main/scala/zio/Promise.scala

```diff
@@ -201,33 +203,37 @@ final class Promise[E, A] private (private val state: AtomicReference[State[E, A
   }
 
   private[zio] def unsafeDone(io: IO[E, A]): Unit = {
-    var retry: Boolean                 = true
-    var joiners: List[IO[E, A] => Any] = null
+    var retry: Boolean                     = true
+    var joiners: Iterable[IO[E, A] => Any] = null
 
     while (retry) {
       val oldState = state.get
 
       val newState = oldState match {
-        case Pending(js) =>
-          joiners = js
+        case Pending(js, _) =>
+          joiners = js.values
```

Comment:
Should you collect this at all or store the map directly and then `foreachValue` in the end?